### PR TITLE
Fix bug where gltfPipeline was not writingSources in all cases 

### PR DESCRIPTION
**Pull request by JoshuaStorm**
Addresses #93 

`gltfPipeline` now writeSources properly when using `processJSON` or `processFile`. Had to move some logic around which caused a bit more copy-pasted code instead of just reusing the `processJSON` method as we were before, but it was only a few lines so it should be fine.

### CODE REVIEW

File: specs/lib/gltfPipelineSpec.js [hunk 1]
```diff
@@ -114,13 +114,47 @@ describe('gltfPipeline', function() {
...
+            callback();
+        });
+        var options = {
+            createDirectory : false,
+            embed : false
+        };
+        readGltf(gltfPath, options, function(gltf) {
+            processJSON(gltf, outputPath, options, function() {
+                process.nextTick(function() {
+                    expect(spy.calls.count()).toEqual(4);
```
Comment (lasalvavida) [inline specs/lib/gltfPipelineSpec.js]: @JoshuaStorm Can you add a check in these tests making sure that source actually does get updated? Maybe make a change to it while it is in pipeline extras then decode the uri after and compare.

Comment (lasalvavida) [inline specs/lib/gltfPipelineSpec.js]: Actually disregard this, it isn't necessary.

**Comment by lasalvavida** [comment]
This looks good to me. @lilleyse is there a good use case to test this with to make sure it works properly?

File: lib/gltfPipeline.js [hunk 2]
```diff
@@ -26,13 +29,24 @@ module.exports = {
...
 };
 
-function processJSON(gltf, options, callback) {
+function writeSources(gltf, outputPath, options, callback) {
+    var embed = options.embed;
+    var embedImage = options.embedImage;
+    var basePath = path.dirname(outputPath);
+    async.each(['buffers', 'images', 'shaders'], function(name) {
+        writeSource(gltf, basePath, name, embed, embedImage, callback);
+    });
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: You need a callback for when the async finishes, and instead you should send in `asyncCallback` to `writeSource`. Check out how `writeGltf` handles this.

File: lib/gltfPipeline.js [hunk 3]
```diff
@@ -26,13 +29,24 @@ module.exports = {
...
+}
+
+function processJSON(gltf, outputPath, options, callback) {
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: `outputPath` doesn't really make sense as an argument for `processJSON` since all we want is JSON back. It looks like it was added so that `writeSources` would work, but instead `writeSources` could just always set `embed` to true.

File: lib/gltfPipeline.js [hunk 4]
```diff
@@ -26,13 +29,24 @@ module.exports = {
...
     addPipelineExtras(gltf);
     loadGltfUris(gltf, options, function(err, gltf) {
         if (err) {
             throw err;
         }
-        processJSONWithExtras(gltf, options, callback);
+        processJSONWithExtras(gltf, options, function(gltf) {
+            writeSources(gltf, outputPath, options, callback(gltf));
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: The callbacks need to be fixed a bit. Normally you could just pass in `callback`, not `callback(gltf)`, but in this case `writeSource` doesn't supply the gltf back, so here you may want:

```javascript
processJSONWithExtras(gltf, options, function(gltf) {
    writeSources(gltf, outputPath, options, function() {
        callback(gltf);
    });
});
```

File: lib/gltfPipeline.js [hunk 5]
```diff
@@ -63,10 +77,10 @@ function processJSONWithExtras(gltfWithExtras, options, callback) {
     }
 }
 
-function processFile(inputPath, options, callback) {
+function processFile(inputPath, outputPath, options, callback) {
     readGltf(inputPath, options, function(gltf) {
         processJSONWithExtras(gltf, options, function(gltf) {
-            callback(gltf);
+            writeSources(gltf, outputPath, options, callback(gltf));
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: Same note here.

File: lib/gltfPipeline.js [hunk 6]
```diff
@@ -84,17 +98,24 @@ function writeFile(gltf, outputPath, options, callback) {
...
-    processJSON(gltf, options, function(gltf) {
-        writeFile(gltf, outputPath, options, callback);
+    addPipelineExtras(gltf);
+    loadGltfUris(gltf, options, function(err, gltf) {
+        if (err) {
+            throw err;
+        }
+        processJSONWithExtras(gltf, options, function(gltf) {
+            writeFile(gltf, outputPath, options, callback);
+        });
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: Why the change here and in `processFileToDisk`?

File: lib/gltfPipeline.js [hunk 7]
```diff
@@ -84,17 +98,24 @@ function writeFile(gltf, outputPath, options, callback) {
...
 function processFileToDisk(inputPath, outputPath, options, callback) {
-    processFile(inputPath, options, function(gltf) {
-        writeFile(gltf, outputPath, options, callback);
+    readGltf(inputPath, options, function(gltf) {
+        processJSONWithExtras(gltf, options, function(gltf) {
+            writeFile(gltf, outputPath, options, callback);
+        });
     });
 }
-
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: Keep this line.

File: specs/lib/gltfPipelineSpec.js [hunk 8]
```diff
@@ -114,13 +114,47 @@ describe('gltfPipeline', function() {
         });
     });
 
+    it('will write sources from JSON', function(done) {
```
Comment (lilleyse) [inline specs/lib/gltfPipelineSpec.js]: This test and the other may not be needed based on the comments above.

File: specs/lib/gltfPipelineSpec.js (hunk 1 above)
Comment (lilleyse) [inline specs/lib/gltfPipelineSpec.js]: I think this is the right idea for testing this.

**Comment by lilleyse** [comment]
> This looks good to me. @lilleyse is there a good use case to test this with to make sure it works properly?

Mainly just your other comment, running a gltf through `processJson` or `processFile` and seeing that its data changed.

File: lib/gltfPipeline.js (hunk 3 above)
Comment (JoshuaStorm) [inline lib/gltfPipeline.js]: I may have misunderstood how we wanted to resolve this issue. I thought we wanted the capability to output non-embedded files when using `processJSON` and `processFile`. If we are fine with them always being embedded, then that does simplify the solution (RE: your comment below - the only reason I refactored `processJSONToDisk` and `processFileToDisk` was to allow non-embedded).

File: lib/gltfPipeline.js (hunk 4 above)
Comment (JoshuaStorm) [inline lib/gltfPipeline.js]: I was wondering why the syntax I used above wouldn't work: Can a function in JS not be passed as an arguments with arguments in the function itself?

File: lib/gltfPipeline.js (hunk 3 above)
Comment (lilleyse) [inline lib/gltfPipeline.js]: Really its not a perfect solution either way, but I think its more user friendly to not write the files. Also I probably just didn't think through it enough when I created the issue.

File: lib/gltfPipeline.js (hunk 4 above)
Comment (lilleyse) [inline lib/gltfPipeline.js]: What will happen is the result of the `callback(gltf)` will be sent to `writeSources`, which could be anything and is probably not itself a callback.

**Comment by JoshuaStorm** [comment]
Hit a bit of a snag, but this good for another look now.

**Comment by likangning93** [comment]
Does this ensure that `writeSources` also happens for the data flow that ends up going through `writeBinaryGltf`? I think that was also an issue, but I might have missed something.

File: lib/gltfPipeline.js (hunk 6 above)
Comment (lilleyse) [inline lib/gltfPipeline.js]: This can change back to its original version now, right?

File: lib/writeSource.js [hunk 9]
```diff
@@ -24,7 +24,7 @@ function writeSource(gltf, basePath, name, embed, embedImage, callback) {
                 var extension = object.extras._pipeline.extension;
 
                 // Write the source object as a data or file uri depending on the embed flag
-                if (embed && (embedImage || name !== 'images')) {
+                if (embed && (embedImage || name !== 'images') || basePath === undefined) {
```
Comment (lilleyse) [inline lib/writeSource.js]: Use `Cesium.defined` here.

File: specs/lib/gltfPipelineSpec.js [hunk 10]
```diff
@@ -104,22 +104,46 @@ describe('gltfPipeline', function() {
             callback();
         });
         var options = {
-            createDirectory : false
+            createDirectory : false,
```
Comment (lilleyse) [inline specs/lib/gltfPipelineSpec.js]: Remove comma.

File: specs/lib/gltfPipelineSpec.js [hunk 11]
```diff
@@ -104,22 +104,46 @@ describe('gltfPipeline', function() {
...
+    });
+
+    it('will write sources from JSON', function(done) {
+        var options = {};
+        readGltf(gltfPath, options, function (gltf) {
+            var initialSource = gltf['buffers'].CesiumTexturedBoxTest.extras._pipeline.source;
+            processJSON(gltf, options, function () {
+                var finalSource = gltf['buffers'].CesiumTexturedBoxTest.extras._pipeline.source;
+                expect(initialSource).not.toEqual(finalSource);
+                done();
```
Comment (lilleyse) [inline specs/lib/gltfPipelineSpec.js]: For these two tests it's more accurate to compare the data uris since we are checking that it wrote the data back to those.

File: specs/lib/gltfPipelineSpec.js [hunk 12]
```diff
@@ -104,22 +104,46 @@ describe('gltfPipeline', function() {
...
+                expect(initialSource).not.toEqual(finalSource);
                 done();
             });
         });
     });
 
     it('will add image processing extras if this is a pipeline with image processing', function(done) {
         var options = {
-            imageProcess: true
+            imageProcess : true,
```
Comment (lilleyse) [inline specs/lib/gltfPipelineSpec.js]: Remove comma.

**Comment by lilleyse** [comment]
> Does this ensure that writeSources also happens for the data flow that ends up going through writeBinaryGltf? I think that was also an issue, but I might have missed something.

`writeBinaryGltf` and `writeFile` should both handle writing `extras._pipeline.source` back to the gltf, the changes in this PR only deal with the function that return a gltf as JSON.

**Comment by JoshuaStorm** [comment]
Updated.

> writeBinaryGltf and writeFile should both handle writing extras._pipeline.source back to the gltf, the changes in this PR only deal with the function that return a gltf as JSON.

Is there currently an issue with `writeBinaryGltf` not having a writeSource step? I could look at that issue next.

**Comment by lilleyse** [comment]
> Is there currently an issue with writeBinaryGltf not having a writeSource step? I could look at that issue next.

No issue with that, it's working fine for me.

File: lib/gltfPipeline.js (hunk 6 above)
Comment (lilleyse) [inline lib/gltfPipeline.js]: Never mind, these functions can't use `processJson` anymore due to the recent changes.

File: specs/lib/gltfPipelineSpec.js [hunk 13]
```diff
@@ -107,19 +107,43 @@ describe('gltfPipeline', function() {
...
+        });
+    });
+
+    it('will write sources from JSON', function(done) {
+        var options = {};
+        readGltf(gltfPath, options, function (gltf) {
+            var initialUri = gltf['buffers'].CesiumTexturedBoxTest.uri;
+            processJSON(gltf, options, function () {
+                var finalUri = gltf['buffers'].CesiumTexturedBoxTest.uri;
+                expect(initialUri).not.toEqual(finalUri);
```
Comment (lilleyse) [inline specs/lib/gltfPipelineSpec.js]: The initial uri will be a filepath to the .bin file, so the comparison will always pass. A quick fix for these 2 tests is to read from `gltfEmbeddedPath` instead of `gltfPath`.

**Comment by JoshuaStorm** [comment]
These are good for review whenever you're ready

**Comment by lilleyse** [comment]
Looks good to me!
